### DISCUSSION TITLE
Generating patches?

### DISCUSSION THREAD

**Discussion by jeroen-mostert** [lamikr/rocm_sdk_builder#85] [Q&A] [answered]
@lamikr What's the command or tooling you use to generate patches? I'd like to make it so my patches match the format used in the repo, but despite fiddling around with many ways to make `git format-patch` spit out something, it never produces exactly. I assume you're not hand-editing the files to include the description, for example. Or maybe you are, would be good to know. :P

**Comment by lamikr** [comment, marked as answer]
No, I just use git format-patch, git commit --amend, git cherry-pick, patch -p1 < patch-name.patch and and other pretty regular commands. I do not like git merge commits, so I always try to make sure patches apply cleanly on top of the latest commit and therefore I use "git rebase and merge" strategy on github.  I think I also once tested that squashing of multiple patches to one also worked.

This is my common sequence for example to latest deepspeed patches.
Note that commands 1 and 2 can be compined to be just (git commit -s) but I quite
ofter use amend command to edit later the description part or tittle to fix typos.

```
1) "git commit -s -m "tittle of patch"
2) git commit --amend
     add some description and then the
     fixes: url line if there is bug open
3) git format-patch <tag-of-first-default-version-without-patches> or
     git format-patch HEAD~4 for example
4) rm -rf ../../patches/rocm-6.1.2/DeepSpeed/00*
5) mv 00* ../../patches/rocm-6.1.2/DeepSpeed
6) cd ../../patches/rocm-6.1.2/DeepSpeed
7) git status .
8) git add * 
(and git rm patch name sometimes if something has been deleted on that dir)
9) git commit -s 
(and put often same commit title than to real patch)
10) git push origin HEAD:wip/rocm_sdk_release_bg74

```
If I need to make changes to patch, I usually at some point to compine them still to one patch if that makes logically sense. So if the first patch has been used to modify the DeepSpeed/build_and_install.sh for example and I need to make changes to it. Now I just created a new patch on bg74 pull request to modify the file created by that patch but at some point, I will propably compine them to single patch.

```
git format-patch HEAD~3
git checkout HEAD~3
patch -p1 < 0003-mod.patch
rm -f 0003-mod.patch
git add build_and_install_rocm.sh
git commit --amend

git am 0001.patch
git am 0002.patch (or git am 00*)
git format-patch ...
mv 00* ../../patches

```
I do not think I have anything special on gitconfig either. (Sometimes I have had there commands to autoclean whitespaces, line-endings, etc. but not sure would they start cleaning whitespaces from patches that I commit on another patch...)

```
[color]
	ui = auto
[filter "lfs"]
	process = git-lfs filter-process
	required = true
	clean = git-lfs clean -- %f
	smudge = git-lfs smudge -- %f
[init]
	defaultBranch = master
[advice]
	detachedHead = false

```

**Comment by jeroen-mostert** [comment]
Thanks. I'm not used to multiline commit messages so I was futzing around with the branch description, but there's no way to get the necessary output that way. For some reason I also didn't consider regenerating all the patches from the source in the `rocm_sdk` directory, I was generating my (one) patch directly from the unmodified repo. In hindsight the workflow is obvious... but that's why they call it hindsight, I suppose.
